### ISSUE TITLE
Composed primary keys

### ISSUE THREAD

**Issue by stylesuxx** [cswertz/lerndorf#24]
I have now reached the point where I am adding the Models for the Entities with composed primary keys for versioning. I have been thinking about this and I am not a fan of using composed primary keys. It makes relations quite a pain, also every table referencing a table with composed primary keys needs n fields to match all members of the composed primary key.

It also makes coding a bit painful - Instead of addressing an entity like so:

    /api/entity/:id

It needs to be addressed like so:

    /api/entity/:id/:time

meaning the queries behind the endpoints also get quite a bit more complicated.

Thus I suggest the following:
Every entity that needs versioning gets a field named **root**. **root** is a reference to the original entity. Further I would add a **parent** field, this would reference the previous iteration of the versioned entity which would makes it very easy to navigate through the history of an entity.

Additionally I would add a composite unique constraint on **root** and **createdAt**.

This way we do not lose any benefits of the composed primary key - we can still easily query all versions of one entity by querying the **root** field (and sorting by *createdAt*), with the parent field we can additionally easily *navigate back in time*. The single primary key makes relations and endpoints a lot simpler.

@cswertz if you do not see any downside to this, I would adjust the design accordingly and move on with creating the rest of the models/relations.

**Comment by cswertz** [comment]
OK - let me think. You suggest a backward linked list. In that case, an ID as unique key (the existing ones are not unique but can be changed) would be needed too I guess, since considerably knowledge units can be part or courses later on. And in courses the specific version that has been added to the course needs to be stored. It will also often be necessary to look up if there is a newer version (to show people that there is a newer version). Looking this up might become expensive with a backward linked list. Maybe a forward linked list would make this easier? I assume that looking up older versions, which becomes more expensive in that case, will not happen to often. Thus a "child" field instead of a "parent" field might be an idea. Similar for a root field. As far as I understood this would make it easier to look up the first version. But honestly - who wants to see first versions? We want to see new versions and updates - thus the child version will be more helpful and a root field is not really necessary. Or do I overlook something here?

Just as a side remark: I've used a double linked list for course sequences in an earlier version. And that was a pain for me. I always mixed the references up. Well - I tried to repress that. But ghosts always come back :-)

**Comment by stylesuxx** [comment]
Yes, a backward linked list.
Having a parent field would keep the updates to a minimum. When a new version is added, only the created entity needs to be touched. With a child field the previous entity has to be updated too.

The root field would allow to get updates with a simple query:

    select * from xxx where rootId=current.rootId and createdAt>current.createdAt order by createdAt

Similarly getting the latest version:

    select * from xxx where rootId=current.rootId and createdAt>current.createdAt order by createdAt DESC limit 1

rootId and createdAt are composed unique and could thus be easily used as an index, thus the newly added primary key would just be a surrogate key.

With a child field (and without a root field), we would also need to traverse to the leaf of the linked list to get the most recent version. The parent field was just to make it easy to compare with the last version. Technically it is not needed at all, since we could do a similar query to the latest version for the previous version:

    select * from xxx where rootId=current.rootId and createdAt<current.createdAt order by createdAt DESC limit 1

### To summarize
We basically still keep the composed primary key, but instead of being a key, it is *just* a composed unique constraint and an index. The new primary key is a surrogate key, it would make referencing and handling the entity in source easier.

Parent and child field are not needed at all. Parent would just be for convenience to compare to the last version but can also be replaced with a simple query.

**Comment by cswertz** [comment]
OK. Updates are seldom (like in the wikipedia: text is often read, but hardly changed). Thus the additional costs of an child field are not very relevant. But we are not always showing the current version, but often older ones (in courses). In that case, a hint that a new version is available would be good. And to find that, we would have to look it up again with a backward list. With a forward list, the new version would be mentioned straight. But well - we have to consider the case that there is more than one version behind the shown one. Thus we would have to look it up again, and thus the direction of the list doesn't matter. So - a backward list is fine. Thank's for the suggestion!
BTW: I'll take care for the rights - problem ASAP.

**Comment by stylesuxx** [comment]
We can also have parent **and** child, then we have the best from both worlds. If you don't mind updating another entity when adding a new one, I am totally OK with it. And yes, the overhead is minimal, it still is overhead though.

But it would easily allow us to naviagate through the history (prev/next) and one query to get the most recent version.

**Comment by cswertz** [comment]
I see no clear resons for one model. Thus: Your choice.
